File: crates-internal/edgemodels/examples/cross_encoder.rs
use edgemodels::cross_encoder::CrossEncoder;
use edgetransformers::models::ModelType;
use edgetransformers::traits::Device;

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    // Load cross-encoder
    let cross_encoder = CrossEncoder::from_registry(
        ModelType::MiniLML6V2CrossEncoder,
        None,
        Device::Cpu,
        None,
    ).await?;
    
    // === Example 1: Score a single pair ===
    let score = cross_encoder.predict(
        "How do I train a neural network?",
        "Neural networks are trained using backpropagation and gradient descent."
    ).await?;
    println!("Relevance score: {:.4}", score);
    
    // === Example 2: Rerank search results ===
    let query = "machine learning algorithms";
    let documents = vec![
        "Machine learning algorithms include decision trees, neural networks, and SVMs.",
        "The weather forecast predicts rain tomorrow.",
        "Deep learning is a subset of machine learning using neural networks.",
        "Cooking recipes for Italian pasta dishes.",
    ];
    
    let ranked_indices = cross_encoder.rerank(query, &documents).await?;
    
    println!("\nReranked results:");
    for (rank, &idx) in ranked_indices.iter().enumerate().take(3) {
        let score = cross_encoder.predict(query, documents[idx]).await?;
        println!("{}. [Score: {:.4}] {}", rank + 1, score, documents[idx]);
    }
    
    Ok(())
}

File: scripts/cross_encoder.py
from sentence_transformers.cross_encoder import CrossEncoder

# Load the model
model = CrossEncoder(
    '/home/olafurj/.cache/kjarni/cross-encoder_ms-marco-MiniLM-L-6-v2/', device='cpu')

# === Example 1: Score a single pair ===
query1 = "How do I train a neural network?"
doc1 = "Neural networks are trained using backpropagation and gradient descent."
score1 = model.predict([(query1, doc1)])
print(f"Relevance score (Python): {score1[0]:.4f}\n")


# === Example 2: Rerank search results ===
query2 = "machine learning algorithms"
documents = [
    "Machine learning algorithms include decision trees, neural networks, and SVMs.",
    "The weather forecast predicts rain tomorrow.",
    "Deep learning is a subset of machine learning using neural networks.",
    "Cooking recipes for Italian pasta dishes.",
]

# Create pairs for batch prediction
pairs = [(query2, doc) for doc in documents]
scores = model.predict(pairs)

# Combine docs and scores and sort
results = sorted(zip(documents, scores), key=lambda x: x[1], reverse=True)

print("Reranked results (Python):")
for doc, score in results:
    print(f"[Score: {score:.4f}] {doc}")